In [24]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import glob # for combining csvs

In [14]:
data_files = glob.glob('./data/*.csv')

In [15]:
df = pd.concat(
  (pd.read_csv(filename) for filename in data_files), 
  ignore_index=True
)

df

tourney_id                tourney_name surface  draw_size tourney_level  \
0      2019-M020                    Brisbane    Hard         32             A   
1      2019-M020                    Brisbane    Hard         32             A   
2      2019-M020                    Brisbane    Hard         32             A   
3      2019-M020                    Brisbane    Hard         32             A   
4      2019-M020                    Brisbane    Hard         32             A   
...          ...                         ...     ...        ...           ...   
73242   2014-605                 Tour Finals    Hard          8             F   
73243   2014-605                 Tour Finals    Hard          8             F   
73244  2014-D015  Davis Cup WG F: FRA vs SUI    Clay          4             D   
73245  2014-D015  Davis Cup WG F: FRA vs SUI    Clay          4             D   
73246  2014-D015  Davis Cup WG F: FRA vs SUI    Clay          4             D   

       tourney_date  match_num  winner_id winner_seed winner_entry  ...  \
0          20181231        300     105453         2.0          NaN  ...   
1          20181231        299     106421         4.0          NaN  ...   
2          20181231        298     105453         2.0          NaN  ...   
3          20181231        297     104542         NaN           PR  ...   
4          20181231        296     106421         4.0          NaN  ...   
...             ...        ...        ...         ...          ...  ...   
73242      20141109        514     104925         1.0          NaN  ...   
73243      20141109        515     104925         1.0          NaN  ...   
73244      20141121          1     104527         NaN          NaN  ...   
73245      20141121          2     104792         NaN          NaN  ...   
73246      20141121          4     103819         NaN          NaN  ...   

      l_1stIn l_1stWon  l_2ndWon l_SvGms  l_bpSaved  l_bpFaced winner_rank  \
0        54.0     34.0      20.0    14.0       10.0       15.0         9.0   
1        52.0     36.0       7.0    10.0       10.0       13.0        16.0   
2        27.0     15.0       6.0     8.0        1.0        5.0         9.0   
3        60.0     38.0       9.0    11.0        4.0        6.0       239.0   
4        56.0     46.0      19.0    15.0        2.0        4.0        16.0   
...       ...      ...       ...     ...        ...        ...         ...   
73242    38.0     29.0       5.0    11.0        3.0        9.0         1.0   
73243     NaN      NaN       NaN     NaN        NaN        NaN         1.0   
73244     NaN      NaN       NaN     NaN        NaN        NaN         4.0   
73245     NaN      NaN       NaN     NaN        NaN        NaN        19.0   
73246     NaN      NaN       NaN     NaN        NaN        NaN         2.0   

      winner_rank_points loser_rank loser_rank_points  
0                 3590.0       16.0            1977.0  
1                 1977.0      239.0             200.0  
2                 3590.0       40.0            1050.0  
3                  200.0       31.0            1298.0  
4                 1977.0       18.0            1855.0  
...                  ...        ...               ...  
73242            10010.0        5.0            4625.0  
73243            10010.0        2.0            8700.0  
73244             5295.0       12.0            2740.0  
73245             1825.0        2.0            9700.0  
73246             9700.0       26.0            1350.0  

[73247 rows x 49 columns]

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 73247 entries, 0 to 73246
Data columns (total 49 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   tourney_id          73247 non-null  object 
 1   tourney_name        73247 non-null  object 
 2   surface             73194 non-null  object 
 3   draw_size           73247 non-null  int64  
 4   tourney_level       73247 non-null  object 
 5   tourney_date        73247 non-null  int64  
 6   match_num           73247 non-null  int64  
 7   winner_id           73247 non-null  int64  
 8   winner_seed         30444 non-null  object 
 9   winner_entry        9229 non-null   object 
 10  winner_name         73247 non-null  object 
 11  winner_hand         73240 non-null  object 
 12  winner_ht           71351 non-null  float64
 13  winner_ioc          73247 non-null  object 
 14  winner_age          73240 non-null  float64
 15  loser_id            73247 non-null  int64  
 16  lose

In [21]:
df = df.sort_values(by='tourney_date')

In [23]:
df[df['winner_rank'].isna()]

tourney_id                  tourney_name surface  \
69918                          2000-891                       Chennai    Hard   
69940                          2000-891                       Chennai    Hard   
69933                          2000-891                       Chennai    Hard   
70277                         2000-D069   Davis Cup G2 QF: GUA vs CUB    Hard   
70284                         2000-D071   Davis Cup G2 QF: ESA vs PAR    Clay   
...                                 ...                           ...     ...   
4201   2024-M-DC-2024-WG2-PO-HKG-ZIM-01  Davis Cup WG2 PO: HKG vs ZIM    Hard   
4206   2024-M-DC-2024-WG2-PO-PAR-MON-01  Davis Cup WG2 PO: PAR vs MON    Clay   
4205   2024-M-DC-2024-WG2-PO-JAM-BAR-01  Davis Cup WG2 PO: JAM vs BAR    Hard   
4181   2024-M-DC-2024-WG1-PO-PAK-IND-01  Davis Cup WG1 PO: PAK vs IND   Grass   
4188   2024-M-DC-2024-WG2-PO-BOL-THA-01  Davis Cup WG2 PO: BOL vs THA    Clay   

       draw_size tourney_level  tourney_date  match_num  winner_id  \
69918         32             A      20000103          3     210013   
69940         32             A      20000103         25     210013   
69933         32             A      20000103         18     210013   
70277          4             D      20000204          2     108582   
70284          4             D      20000204          1     102740   
...          ...           ...           ...        ...        ...   
4201           4             D      20240203          4     212721   
4206           4             D      20240203          1     118808   
4205           4             D      20240203          5     209868   
4181           4             D      20240203          2     105502   
4188           4             D      20240203          4     106262   

      winner_seed winner_entry  ... l_1stIn l_1stWon  l_2ndWon l_SvGms  \
69918         NaN          NaN  ...    17.0      8.0       9.0     7.0   
69940         NaN          NaN  ...    62.0     40.0      13.0    10.0   
69933         NaN          NaN  ...    53.0     34.0       8.0    12.0   
70277         NaN          NaN  ...     NaN      NaN       NaN     NaN   
70284         NaN          NaN  ...     NaN      NaN       NaN     NaN   
...           ...          ...  ...     ...      ...       ...     ...   
4201          NaN          NaN  ...    49.0     35.0      17.0    11.0   
4206          NaN          NaN  ...    31.0     17.0      10.0     9.0   
4205          NaN          NaN  ...    55.0     34.0      12.0    12.0   
4181          NaN          NaN  ...    27.0     23.0      16.0    10.0   
4188          NaN          NaN  ...    38.0     28.0      13.0     9.0   

       l_bpSaved  l_bpFaced winner_rank winner_rank_points loser_rank  \
69918        1.0        5.0         NaN                NaN      121.0   
69940       11.0       13.0         NaN                NaN      115.0   
69933        4.0        8.0         NaN                NaN       96.0   
70277        NaN        NaN         NaN                NaN        NaN   
70284        NaN        NaN         NaN                NaN        NaN   
...          ...        ...         ...                ...        ...   
4201         0.0        2.0         NaN                NaN        NaN   
4206         3.0        7.0         NaN                NaN        NaN   
4205         3.0        8.0         NaN                NaN        NaN   
4181         3.0        5.0         NaN                NaN        NaN   
4188         6.0        9.0         NaN                NaN      938.0   

      loser_rank_points  
69918             385.0  
69940             399.0  
69933             492.0  
70277               NaN  
70284               NaN  
...                 ...  
4201                NaN  
4206                NaN  
4205                NaN  
4181                NaN  
4188               16.0  

[566 rows x 49 columns]

In [25]:
# EDA

In [26]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 73247 entries, 68837 to 4090
Data columns (total 49 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   tourney_id          73247 non-null  object 
 1   tourney_name        73247 non-null  object 
 2   surface             73194 non-null  object 
 3   draw_size           73247 non-null  int64  
 4   tourney_level       73247 non-null  object 
 5   tourney_date        73247 non-null  int64  
 6   match_num           73247 non-null  int64  
 7   winner_id           73247 non-null  int64  
 8   winner_seed         30444 non-null  object 
 9   winner_entry        9229 non-null   object 
 10  winner_name         73247 non-null  object 
 11  winner_hand         73240 non-null  object 
 12  winner_ht           71351 non-null  float64
 13  winner_ioc          73247 non-null  object 
 14  winner_age          73240 non-null  float64
 15  loser_id            73247 non-null  int64  
 16  loser_

In [28]:
df.drop_duplicates(inplace=True)

In [29]:
df.info() # no duplicates!

<class 'pandas.core.frame.DataFrame'>
Index: 73247 entries, 68837 to 4090
Data columns (total 49 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   tourney_id          73247 non-null  object 
 1   tourney_name        73247 non-null  object 
 2   surface             73194 non-null  object 
 3   draw_size           73247 non-null  int64  
 4   tourney_level       73247 non-null  object 
 5   tourney_date        73247 non-null  int64  
 6   match_num           73247 non-null  int64  
 7   winner_id           73247 non-null  int64  
 8   winner_seed         30444 non-null  object 
 9   winner_entry        9229 non-null   object 
 10  winner_name         73247 non-null  object 
 11  winner_hand         73240 non-null  object 
 12  winner_ht           71351 non-null  float64
 13  winner_ioc          73247 non-null  object 
 14  winner_age          73240 non-null  float64
 15  loser_id            73247 non-null  int64  
 16  loser_

In [30]:
# checking for nulls

In [31]:
df.isna().sum(axis=0).sort_values(ascending=False)

winner_entry          64018
loser_entry           58240
loser_seed            56389
winner_seed           42803
minutes                7938
l_1stIn                6469
l_svpt                 6469
l_ace                  6469
l_df                   6469
w_1stWon               6469
w_ace                  6469
w_df                   6469
w_svpt                 6469
w_2ndWon               6469
w_1stIn                6469
l_1stWon               6469
l_2ndWon               6469
l_bpFaced              6469
l_bpSaved              6469
w_bpFaced              6469
w_bpSaved              6469
w_SvGms                6468
l_SvGms                6468
loser_ht               3621
winner_ht              1896
loser_rank             1453
loser_rank_points      1453
winner_rank             566
winner_rank_points      566
surface                  53
loser_hand               42
loser_age                12
winner_age                7
winner_hand               7
winner_ioc                0
loser_id            

In [32]:
missing_stats = df.isna().sum(axis=0).sort_values(ascending=False)/df.shape[0]
missing_stats

winner_entry          0.874002
loser_entry           0.795118
loser_seed            0.769847
winner_seed           0.584365
minutes               0.108373
l_1stIn               0.088318
l_svpt                0.088318
l_ace                 0.088318
l_df                  0.088318
w_1stWon              0.088318
w_ace                 0.088318
w_df                  0.088318
w_svpt                0.088318
w_2ndWon              0.088318
w_1stIn               0.088318
l_1stWon              0.088318
l_2ndWon              0.088318
l_bpFaced             0.088318
l_bpSaved             0.088318
w_bpFaced             0.088318
w_bpSaved             0.088318
w_SvGms               0.088304
l_SvGms               0.088304
loser_ht              0.049435
winner_ht             0.025885
loser_rank            0.019837
loser_rank_points     0.019837
winner_rank           0.007727
winner_rank_points    0.007727
surface               0.000724
loser_hand            0.000573
loser_age             0.000164
winner_a

Winner entry and loser entry are expected to be high. Those columns are there to indicate if a player has come from the qualifications for the tournament or is a wildcard.  This is generally not the case so it makes sense that it's high.

The two columns that I care about for my initial EDA are winner_rank and loser_rank. I want to build my baseline model off of those parameters so any columns where that is null will have to be removed.

In [36]:
df[df['winner_rank'].isna()].info()

<class 'pandas.core.frame.DataFrame'>
Index: 566 entries, 69918 to 4188
Data columns (total 49 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   tourney_id          566 non-null    object 
 1   tourney_name        566 non-null    object 
 2   surface             558 non-null    object 
 3   draw_size           566 non-null    int64  
 4   tourney_level       566 non-null    object 
 5   tourney_date        566 non-null    int64  
 6   match_num           566 non-null    int64  
 7   winner_id           566 non-null    int64  
 8   winner_seed         2 non-null      object 
 9   winner_entry        24 non-null     object 
 10  winner_name         566 non-null    object 
 11  winner_hand         559 non-null    object 
 12  winner_ht           161 non-null    float64
 13  winner_ioc          566 non-null    object 
 14  winner_age          559 non-null    float64
 15  loser_id            566 non-null    int64  
 16  loser_se

In [42]:
non_null_fltr1 = df['winner_rank'].notna()
non_null_fltr2 = df['loser_rank'].notna()

# remove those rows because they're small portion of data
df = df.loc[non_null_fltr1 | non_null_fltr2, :]

In [45]:
df['winner_rank'].value_counts(dropna=False)
df['loser_rank'].value_counts(dropna=False)

# confirmed removed

loser_rank
44.0      596
46.0      592
39.0      589
37.0      588
40.0      586
         ... 
1016.0      1
1561.0      1
968.0       1
1576.0      1
913.0       1
Name: count, Length: 1344, dtype: int64

Now in addition to removing the nulls for those columns, I want to restrict the anaysis to only matches that are Grand Slams or Masters 1000s. These are tournaments that are worth the most points and therefore tend to bring out the best in players. If I want to add more tournaments later I can do so too.

In [47]:
df['tourney_level'].value_counts()

tourney_level
A    39881
M    13726
G    12159
D     5328
F      407
Name: count, dtype: int64

In [62]:
# update dtype to str


tourney_fltr = (df['tourney_level'] == 'M') | (df['tourney_level'] == 'G')

# remove those rows because they're small portion of data
df = df.loc[tourney_fltr, :]

In [65]:
df['tourney_level'].value_counts()

tourney_level
M    13726
G    12159
Name: count, dtype: int64

Now I would like to construct a quick and easy baseline model. For this I just want to see the results if I were to pick the higher ranked player everytime.

In [70]:
higher_rnk_fltr = df['winner_rank'] < df['loser_rank']

df[higher_rnk_fltr].shape[0] / df.shape[0]

0.6761444852231022

This is my baseline model that I will measure the performance of my model against. For that simple model, the winner is correctly picked roughly 68% of the time. 